src/main/java/org/apache/commons/collections4/bloomfilter/SimpleBloomFilter.java

```diff
@@ -0,0 +1,245 @@
+     * @param shape The shape for the filter.
+     * @param indices the IndexProducer to initialize the filter with.
+     * @throws IllegalArgumentException if producer sends illegal value.
+     */
+    public SimpleBloomFilter(final Shape shape, IndexProducer indices) {
+        this(shape);
+        Objects.requireNonNull(indices, "indices");
+        mergeInPlace(indices);
+    }
+
+    /**
+     * Creates a populated instance.
+     * @param shape The shape for the filter.
+     * @param bitMaps the BitMapProducer to initialize the filter with.
+     * @throws IllegalArgumentException if the producer returns too many or too few bit maps.
+     */
+    public SimpleBloomFilter(final Shape shape, BitMapProducer bitMaps) {
+        this(shape);
+        Objects.requireNonNull(bitMaps, "bitMaps");
+        mergeInPlace(bitMaps);
+    }
+
+    /**
+     * Copy constructor for {@code copy()} use.
+     * @param source
+     */
+    private SimpleBloomFilter(SimpleBloomFilter source) {
+        this.shape = source.shape;
+        this.bitMap = source.bitMap.clone();
+        this.cardinality = source.cardinality;
+    }
+
+    @Override
+    public long[] asBitMapArray() {
+        return Arrays.copyOf(bitMap, bitMap.length);
+    }
+
+    @Override
+    public boolean forEachBitMapPair(BitMapProducer other, LongBiPredicate func) {
+        CountingLongPredicate p = new CountingLongPredicate(bitMap, func);
+        return other.forEachBitMap(p) && p.forEachRemaining();
+    }
+
+    @Override
+    public SimpleBloomFilter copy() {
+        return new SimpleBloomFilter(this);
+    }
+
+    /**
+     * Performs a merge in place using an IndexProducer.
+     * @param indexProducer the IndexProducer to merge from.
+     * @throws IllegalArgumentException if producer sends illegal value.
+     */
+    private void mergeInPlace(IndexProducer indexProducer) {
+        indexProducer.forEachIndex(idx -> {
+            if (idx < 0 || idx >= shape.getNumberOfBits()) {
+                throw new IllegalArgumentException(String.format(
+                        "IndexProducer should only send values in the range[0,%s]", shape.getNumberOfBits() - 1));
+            }
+            BitMap.set(bitMap, idx);
+            return true;
+        });
+        cardinality = -1;
+    }
+
+    /**
+     * Performs a merge in place using an BitMapProducer.
+     * @param bitMapProducer the BitMapProducer to merge from.
+     * @throws IllegalArgumentException if producer sends illegal value.
+     */
+    private void mergeInPlace(BitMapProducer bitMapProducer) {
+        try {
+            int[] idx = new int[1];
+            bitMapProducer.forEachBitMap(value -> {
+                bitMap[idx[0]++] |= value;
+                return true;
+            });
+            // idx[0] will be limit+1 so decrement it
+            idx[0]--;
+            int idxLimit = BitMap.getLongIndex(shape.getNumberOfBits());
+            if (idxLimit < idx[0]) {
```

Comment:
There is no coverage of this branch. It is not possible. If this were true the bitMap[idx[0]] in the main forEach loop would have thrown an array index out of bound exception. I think this can be removed. The final check to determine if the final upper bitmap does not set incorrect bits is valid.

Note: Even though the check is made that the upper bits have been set correctly, if they have not the exception is raised and the filter now has incorrect bits for the rest of its lifetime, i.e. no recovery is made and no flag is present in the filter to indicate the bits are bad. So you could then merge this into another filter and get the same exception, or record the bits to file and have bad bits in the recording, etc.

The same 'bug' is present in the SparseBloomFilter. It will raise an exception after merge if the TreeSet first or last position are outside the shape but not correct it.

I think these filters should attempt to clear any invalid bits before throwing exceptions. This makes them act as if they are a plain `int[]` of size `nbits` containing 0 or 1s. It should not be possible to put an index outside the range of the shape in the filter.